<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>valr overview • valr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="valr overview">
<meta property="og:description" content="valr">
<meta property="og:image" content="https://rnabioco.github.io/valr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-54095998-3"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-54095998-3');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">valr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/valr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/benchmarks.html">valr Benchmarks</a>
    </li>
    <li>
      <a href="../articles/interval-stats.html">Interval statistics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/valr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>valr overview</h1>
            
            <h4 class="date">2020-03-24</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/valr/blob/master/vignettes/valr.Rmd"><code>vignettes/valr.Rmd</code></a></small>
      <div class="hidden name"><code>valr.Rmd</code></div>

    </div>

    
    
<div id="why-valr" class="section level2">
<h2 class="hasAnchor">
<a href="#why-valr" class="anchor"></a>Why <code>valr</code>?</h2>
<p><strong>Why another tool set for interval manipulations?</strong> There are several <a href="#related_work">other software packages</a> available for genome interval analysis. However, based on our experiences teaching genome analysis, we were motivated to develop a toolset that:</p>
<ul>
<li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the “modern R” syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code><a href="https://dplyr.tidyverse.org/reference/reexports.html">%&gt;%</a></code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul>
<p><code>valr</code> can currently be used for analysis of pre-processed data in BED and related formats. We plan to support BAM and VCF files soon via tabix indexes.</p>
<div id="familiar-tools-natively-in-r" class="section level3">
<h3 class="hasAnchor">
<a href="#familiar-tools-natively-in-r" class="anchor"></a>Familiar tools, natively in R</h3>
<p>The functions in <code>valr</code> have similar names to their <code>BEDtools</code> counterparts, and so will be familiar to users coming from the <code>BEDtools</code> suite. Similar to <a href="https://daler.github.io/pybedtools/#why-pybedtools"><code>pybedtools</code></a>, <code>valr</code> has a terse syntax:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">valr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)

<span class="no">snps</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="fu"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.snps147.chr22.bed.gz'</span>), <span class="kw">n_fields</span> <span class="kw">=</span> <span class="fl">6</span>)
<span class="no">genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="fu"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>), <span class="kw">n_fields</span> <span class="kw">=</span> <span class="fl">6</span>)

<span class="co"># find snps in intergenic regions</span>
<span class="no">intergenic</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(<span class="no">snps</span>, <span class="no">genes</span>)
<span class="co"># distance from intergenic snps to nearest gene</span>
<span class="no">nearby</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/bed_closest.html">bed_closest</a></span>(<span class="no">intergenic</span>, <span class="no">genes</span>)

<span class="no">nearby</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">starts_with</a></span>(<span class="st">'name'</span>), <span class="no">.overlap</span>, <span class="no">.dist</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">.dist</span>) <span class="kw">&lt;</span> <span class="fl">1000</span>)
<span class="co">#&gt; # A tibble: 285 x 4</span>
<span class="co">#&gt;    name.x      name.y            .overlap .dist</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;                &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt;  1 rs2261631   P704P                    0  -268</span>
<span class="co">#&gt;  2 rs570770556 POTEH                    0  -913</span>
<span class="co">#&gt;  3 rs538163832 POTEH                    0  -953</span>
<span class="co">#&gt;  4 rs9606135   TPTEP1                   0  -422</span>
<span class="co">#&gt;  5 rs11912392  ANKRD62P1-PARP4P3        0   105</span>
<span class="co">#&gt;  6 rs8136454   BC038197                 0   356</span>
<span class="co">#&gt;  7 rs5992556   XKR3                     0  -456</span>
<span class="co">#&gt;  8 rs114101676 GAB4                     0   474</span>
<span class="co">#&gt;  9 rs62236167  CECR7                    0   262</span>
<span class="co">#&gt; 10 rs5747023   CECR1                    0  -387</span>
<span class="co">#&gt; # … with 275 more rows</span></pre></body></html></div>
</div>
<div id="input-data" class="section level3">
<h3 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input data</h3>
<p><code>valr</code> assigns common column names to facilitate comparisons between tbls. All tbls will have <code>chrom</code>, <code>start</code>, and <code>end</code> columns, and some tbls from multi-column formats will have additional pre-determined column names. See the <code><a href="../reference/read_bed.html">read_bed()</a></code> documentation for details.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">bed_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">"3fields.bed.gz"</span>)
<span class="fu"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="no">bed_file</span>) <span class="co"># accepts filepaths or URLs</span>
<span class="co">#&gt; # A tibble: 10 x 3</span>
<span class="co">#&gt;    chrom  start    end</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt;  1 chr1   11873  14409</span>
<span class="co">#&gt;  2 chr1   14361  19759</span>
<span class="co">#&gt;  3 chr1   14406  29370</span>
<span class="co">#&gt;  4 chr1   34610  36081</span>
<span class="co">#&gt;  5 chr1   69090  70008</span>
<span class="co">#&gt;  6 chr1  134772 140566</span>
<span class="co">#&gt;  7 chr1  321083 321115</span>
<span class="co">#&gt;  8 chr1  321145 321207</span>
<span class="co">#&gt;  9 chr1  322036 326938</span>
<span class="co">#&gt; 10 chr1  327545 328439</span></pre></body></html></div>
<p><code>valr</code> can also operate on BED-like data.frames already constructed in R, provided that columns named <code>chrom</code>, <code>start</code> and <code>end</code> are present. New tbls can also be contructed using <code><a href="../reference/tbl_interval.html">trbl_interval()</a></code>.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">bed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/tbl_interval.html">trbl_interval</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>,  ~<span class="no">end</span>,
  <span class="st">"chr1"</span>, <span class="fl">1657492</span>, <span class="fl">2657492</span>,
  <span class="st">"chr2"</span>, <span class="fl">2501324</span>, <span class="fl">3094650</span>
)

<span class="no">bed</span>
<span class="co">#&gt; # A tibble: 2 x 3</span>
<span class="co">#&gt;   chrom   start     end</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr1  1657492 2657492</span>
<span class="co">#&gt; 2 chr2  2501324 3094650</span></pre></body></html></div>
</div>
<div id="interval-coordinates" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-coordinates" class="anchor"></a>Interval coordinates</h3>
<p><code>valr</code> adheres to the BED <a href="http://genome.ucsc.edu/FAQ/FAQformat#format1">format</a> which specifies that the start position for an interval is zero based and the end position is one-based. The first position in a chromosome is 0. The end position for a chromosome is one position passed the last base, and is not included in the interval. For example:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># a chromosome 100 basepairs in length</span>
<span class="no">chrom</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/tbl_interval.html">trbl_interval</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>,
  <span class="st">"chr1"</span>, <span class="fl">0</span>,      <span class="fl">100</span>
)

<span class="no">chrom</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr1      0   100</span>

<span class="co"># single basepair intervals</span>
<span class="no">bases</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/tbl_interval.html">trbl_interval</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>,
  <span class="st">"chr1"</span>, <span class="fl">0</span>,      <span class="fl">1</span>, <span class="co"># first base of chromosome</span>
  <span class="st">"chr1"</span>, <span class="fl">1</span>,      <span class="fl">2</span>,  <span class="co"># second base of chromosome</span>
  <span class="st">"chr1"</span>, <span class="fl">99</span>,     <span class="fl">100</span> <span class="co"># last base of chromosome</span>
)

<span class="no">bases</span>
<span class="co">#&gt; # A tibble: 3 x 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr1      0     1</span>
<span class="co">#&gt; 2 chr1      1     2</span>
<span class="co">#&gt; 3 chr1     99   100</span></pre></body></html></div>
</div>
<div id="remote-databases" class="section level3">
<h3 class="hasAnchor">
<a href="#remote-databases" class="anchor"></a>Remote databases</h3>
<p>Remote databases can be accessed with <code><a href="../reference/db.html">db_ucsc()</a></code> (to access the UCSC Browser) and <code><a href="../reference/db.html">db_ensembl()</a></code> (to access Ensembl databases).</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># access the `refGene` tbl on the `hg38` assembly.</span>
<span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">RMySQL</span>)) {
  <span class="no">ucsc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/db.html">db_ucsc</a></span>(<span class="st">'hg38'</span>)
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span>(<span class="no">ucsc</span>, <span class="st">'refGene'</span>)
}</pre></body></html></div>
</div>
<div id="visual-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#visual-documentation" class="anchor"></a>Visual documentation</h3>
<p>The <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code> tool illustrates the results of operations in <code>valr</code>, similar to those found in the <code>BEDtools</code> documentation. This glyph shows the result of intersecting <code>x</code> and <code>y</code> intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>,
  <span class="st">'chr1'</span>, <span class="fl">25</span>,     <span class="fl">50</span>,
  <span class="st">'chr1'</span>, <span class="fl">100</span>,    <span class="fl">125</span>
)

<span class="no">y</span> <span class="kw">&lt;-</span> <span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>,
  <span class="st">'chr1'</span>, <span class="fl">30</span>,     <span class="fl">75</span>
)

<span class="fu"><a href="../reference/bed_glyph.html">bed_glyph</a></span>(<span class="fu"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(<span class="no">x</span>, <span class="no">y</span>))</pre></body></html></div>
<p><img src="valr_files/figure-html/intersect_glyph-1.png" width="700" style="display: block; margin: auto;"></p>
<p>And this glyph illustrates <code><a href="../reference/bed_merge.html">bed_merge()</a></code>:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>,
  <span class="st">'chr1'</span>,      <span class="fl">1</span>,      <span class="fl">50</span>,
  <span class="st">'chr1'</span>,      <span class="fl">10</span>,     <span class="fl">75</span>,
  <span class="st">'chr1'</span>,      <span class="fl">100</span>,    <span class="fl">120</span>
)

<span class="fu"><a href="../reference/bed_glyph.html">bed_glyph</a></span>(<span class="fu"><a href="../reference/bed_merge.html">bed_merge</a></span>(<span class="no">x</span>))</pre></body></html></div>
<p><img src="valr_files/figure-html/merge_glyph-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="reproducible-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#reproducible-reports" class="anchor"></a>Reproducible reports</h3>
<p><code>valr</code> can be used in RMarkdown documents to generate reproducible work-flows for data processing. Because <code>valr</code> is reasonably fast (see the <a href="#benchmarks">benchmarks</a>), we now use it in lieu of other tools for exploratory analysis of genomic data sets in R.</p>
<p>Command-line tools like <code>BEDtools</code> and <code>bedops</code> can be used in reproducible workflows (e.g., with <a href="https://bitbucket.org/snakemake/snakemake/wiki/Home"><code>snakemake</code></a>), but it is cumbersome to move from command-line tools to exploratory analysis and plotting software. <a href="https://pythonhosted.org/pybedtools/"><code>pybedtools</code></a> can be used within <code>ipython notebooks</code> to accomplish a similar goal, but others have pointed out <a href="https://www.r-bloggers.com/why-i-dont-like-jupyter-fka-ipython-notebook/">issues with this approach</a>, including clunky version control. Because RMarkdown files are text files, they are readily kept under version control. Moreover, new features in RStudio (e.g. notebook viewing) enable similar functionality to <code>ipython</code>.</p>
</div>
<div id="grouping-data" class="section level3">
<h3 class="hasAnchor">
<a href="#grouping-data" class="anchor"></a>Grouping data</h3>
<p>The <code>group_by</code> function in dplyr can be used to perform fuctions on subsets of single and multiple <code>data_frame</code>s. Functions in <code>valr</code> leverage grouping to enable a variety of comparisons. For example, intervals can be grouped by <code>strand</code> to perform comparisons among intervals on the same strand.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>, ~<span class="no">strand</span>,
  <span class="st">'chr1'</span>, <span class="fl">1</span>,      <span class="fl">100</span>,  <span class="st">'+'</span>,
  <span class="st">'chr1'</span>, <span class="fl">50</span>,     <span class="fl">150</span>,  <span class="st">'+'</span>,
  <span class="st">'chr2'</span>, <span class="fl">100</span>,    <span class="fl">200</span>,  <span class="st">'-'</span>
)

<span class="no">y</span> <span class="kw">&lt;-</span> <span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(
  ~<span class="no">chrom</span>, ~<span class="no">start</span>, ~<span class="no">end</span>, ~<span class="no">strand</span>,
  <span class="st">'chr1'</span>, <span class="fl">50</span>,     <span class="fl">125</span>,  <span class="st">'+'</span>,
  <span class="st">'chr1'</span>, <span class="fl">50</span>,     <span class="fl">150</span>,  <span class="st">'-'</span>,
  <span class="st">'chr2'</span>, <span class="fl">50</span>,     <span class="fl">150</span>,  <span class="st">'+'</span>
)

<span class="co"># intersect tbls by strand</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">x</span>, <span class="no">strand</span>)
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">y</span>, <span class="no">strand</span>)

<span class="fu"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(<span class="no">x</span>, <span class="no">y</span>)
<span class="co">#&gt; # A tibble: 2 x 8</span>
<span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt;</span>
<span class="co">#&gt; 1 chr1        1   100 +             50   125 +              50</span>
<span class="co">#&gt; 2 chr1       50   150 +             50   125 +              75</span></pre></body></html></div>
<p>Comparisons between intervals on opposite strands are done using the <code><a href="../reference/flip_strands.html">flip_strands()</a></code> function:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">x</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">x</span>, <span class="no">strand</span>)

<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/flip_strands.html">flip_strands</a></span>(<span class="no">y</span>)
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">y</span>, <span class="no">strand</span>)

<span class="fu"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(<span class="no">x</span>, <span class="no">y</span>)
<span class="co">#&gt; # A tibble: 3 x 8</span>
<span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt;</span>
<span class="co">#&gt; 1 chr1        1   100 +             50   150 +              50</span>
<span class="co">#&gt; 2 chr1       50   150 +             50   150 +             100</span>
<span class="co">#&gt; 3 chr2      100   200 -             50   150 -              50</span></pre></body></html></div>
<p>Both single set (e.g. <code><a href="../reference/bed_merge.html">bed_merge()</a></code>) and multi set operations will respect groupings in the input intervals.</p>
</div>
<div id="column-specification" class="section level3">
<h3 class="hasAnchor">
<a href="#column-specification" class="anchor"></a>Column specification</h3>
<p>Columns in <code>BEDtools</code> are referred to by position:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># calculate the mean of column 6 for intervals in `b` that overlap with `a`</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="ex">bedtools</span> map -a a.bed -b b.bed -c 6 -o mean</span></code></pre></div>
<p>In <code>valr</code>, columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># calculate the mean and variance for a `value` column</span>
<span class="fu"><a href="../reference/bed_map.html">bed_map</a></span>(<span class="no">a</span>, <span class="no">b</span>, <span class="kw">.mean</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">value</span>), <span class="kw">.var</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">var</a></span>(<span class="no">value</span>))

<span class="co"># report concatenated and max values for merged intervals</span>
<span class="fu"><a href="../reference/bed_merge.html">bed_merge</a></span>(<span class="no">a</span>, <span class="kw">.concat</span> <span class="kw">=</span> <span class="fu"><a href="../reference/bed_map.html">concat</a></span>(<span class="no">value</span>), <span class="kw">.max</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">value</span>))</pre></body></html></div>
</div>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<div id="meta-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#meta-analysis" class="anchor"></a>Meta-analysis</h3>
<p>This demonstration illustrates how to use <code>valr</code> tools to perform a “meta-analysis” of signals relative to genomic features. Here we to analyze the distribution of histone marks surrounding transcription start sites.</p>
<p>First we load libraries and relevant data.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># `valr_example()` identifies the path of example files</span>
<span class="no">bedfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
<span class="no">genomefile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>)
<span class="no">bgfile</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

<span class="no">genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="no">bedfile</span>, <span class="kw">n_fields</span> <span class="kw">=</span> <span class="fl">6</span>)
<span class="no">genome</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="no">genomefile</span>)
<span class="no">y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bedgraph</a></span>(<span class="no">bgfile</span>)</pre></body></html></div>
<p>Then we generate 1 bp intervals to represent transcription start sites (TSSs). We focus on <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> strand genes, but <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> genes are easily accomodated by filtering them and using <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> with <code>reverse</code>d window numbers.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co"># generate 1 bp TSS intervals, `+` strand only</span>
<span class="no">tss</span> <span class="kw">&lt;-</span> <span class="no">genes</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">strand</span> <span class="kw">==</span> <span class="st">'+'</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">end</span> <span class="kw">=</span> <span class="no">start</span> + <span class="fl">1</span>)

<span class="co"># 1000 bp up and downstream</span>
<span class="no">region_size</span> <span class="kw">&lt;-</span> <span class="fl">1000</span>
<span class="co"># 50 bp windows</span>
<span class="no">win_size</span> <span class="kw">&lt;-</span> <span class="fl">50</span>

<span class="co"># add slop to the TSS, break into windows and add a group</span>
<span class="no">x</span> <span class="kw">&lt;-</span> <span class="no">tss</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/bed_slop.html">bed_slop</a></span>(<span class="no">genome</span>, <span class="kw">both</span> <span class="kw">=</span> <span class="no">region_size</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/bed_makewindows.html">bed_makewindows</a></span>(<span class="no">win_size</span>)

<span class="no">x</span>
<span class="co">#&gt; # A tibble: 13,530 x 7</span>
<span class="co">#&gt;    chrom    start      end name      score strand .win_id</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;    &lt;int&gt;</span>
<span class="co">#&gt;  1 chr22 16161065 16161115 LINC00516 3     +            1</span>
<span class="co">#&gt;  2 chr22 16161115 16161165 LINC00516 3     +            2</span>
<span class="co">#&gt;  3 chr22 16161165 16161215 LINC00516 3     +            3</span>
<span class="co">#&gt;  4 chr22 16161215 16161265 LINC00516 3     +            4</span>
<span class="co">#&gt;  5 chr22 16161265 16161315 LINC00516 3     +            5</span>
<span class="co">#&gt;  6 chr22 16161315 16161365 LINC00516 3     +            6</span>
<span class="co">#&gt;  7 chr22 16161365 16161415 LINC00516 3     +            7</span>
<span class="co">#&gt;  8 chr22 16161415 16161465 LINC00516 3     +            8</span>
<span class="co">#&gt;  9 chr22 16161465 16161515 LINC00516 3     +            9</span>
<span class="co">#&gt; 10 chr22 16161515 16161565 LINC00516 3     +           10</span>
<span class="co">#&gt; # … with 13,520 more rows</span></pre></body></html></div>
<p>Now we use the <code>.win_id</code> group with <code><a href="../reference/bed_map.html">bed_map()</a></code> to caluclate a sum by mapping <code>y</code> signals onto the intervals in <code>x</code>. These data are regrouped by <code>.win_id</code> and a summary with <code>mean</code> and <code>sd</code> values is calculated.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># map signals to TSS regions and calculate summary statistics.</span>
<span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/bed_map.html">bed_map</a></span>(<span class="no">x</span>, <span class="no">y</span>, <span class="kw">win_sum</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">value</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(<span class="no">.win_id</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="kw">win_mean</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">win_sum</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
            <span class="kw">win_sd</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(<span class="no">win_sum</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>))

<span class="no">res</span>
<span class="co">#&gt; # A tibble: 41 x 3</span>
<span class="co">#&gt;    .win_id win_mean win_sd</span>
<span class="co">#&gt;      &lt;int&gt;    &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1     101.   85.8</span>
<span class="co">#&gt;  2       2     111.   81.1</span>
<span class="co">#&gt;  3       3     123.   99.1</span>
<span class="co">#&gt;  4       4     116.   96.3</span>
<span class="co">#&gt;  5       5     116.  102. </span>
<span class="co">#&gt;  6       6     125.   95.1</span>
<span class="co">#&gt;  7       7     123.   94.4</span>
<span class="co">#&gt;  8       8     128.   91.5</span>
<span class="co">#&gt;  9       9     130.   95.7</span>
<span class="co">#&gt; 10      10     130.   88.8</span>
<span class="co">#&gt; # … with 31 more rows</span></pre></body></html></div>
<p>Finally, these summary statistics are used to construct a plot that illustrates histone density surrounding TSSs.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)

<span class="no">x_labels</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(-<span class="no">region_size</span>, <span class="no">region_size</span>, <span class="kw">by</span> <span class="kw">=</span> <span class="no">win_size</span> * <span class="fl">5</span>)
<span class="no">x_breaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">1</span>, <span class="fl">41</span>, <span class="kw">by</span> <span class="kw">=</span> <span class="fl">5</span>)

<span class="no">sd_limits</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">ymax</span> <span class="kw">=</span> <span class="no">win_mean</span> + <span class="no">win_sd</span>, <span class="kw">ymin</span> <span class="kw">=</span> <span class="no">win_mean</span> - <span class="no">win_sd</span>)

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">res</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">.win_id</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">win_mean</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>(<span class="no">sd_limits</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="kw">labels</span> <span class="kw">=</span> <span class="no">x_labels</span>, <span class="kw">breaks</span> <span class="kw">=</span> <span class="no">x_breaks</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'Position (bp from TSS)'</span>) + <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'Signal'</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">'Human H3K4me3 signal near transcription start sites'</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()</pre></body></html></div>
<p><img src="valr_files/figure-html/plot-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="api" class="section level2">
<h2 class="hasAnchor">
<a href="#api" class="anchor"></a>API</h2>
<p>Function names are similar to their their <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> counterparts, with some additions.</p>
<div id="data-types" class="section level3">
<h3 class="hasAnchor">
<a href="#data-types" class="anchor"></a>Data types</h3>
<ul>
<li>Create interval sets with <code><a href="../reference/tbl_interval.html">tbl_interval()</a></code> and <code><a href="../reference/tbl_genome.html">tbl_genome()</a></code>, which enforce strict column naming.</li>
</ul>
</div>
<div id="reading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-data" class="anchor"></a>Reading data</h3>
<ul>
<li><p>Read BED and related files with <code><a href="../reference/read_bed.html">read_bed()</a></code>, <code><a href="../reference/read_bed.html">read_bed12()</a></code>, <code><a href="../reference/read_bed.html">read_bedgraph()</a></code>, <code><a href="../reference/read_bed.html">read_narrowpeak()</a></code> and <code><a href="../reference/read_bed.html">read_broadpeak()</a></code>.</p></li>
<li><p>Read genome files containing chromosome name and size information with <code><a href="../reference/read_genome.html">read_genome()</a></code>.</p></li>
<li><p>Load VCF files with <code><a href="../reference/read_vcf.html">read_vcf()</a></code>.</p></li>
<li><p>Access remote databases with <code><a href="../reference/db.html">db_ucsc()</a></code> and <code><a href="../reference/db.html">db_ensembl()</a></code>.</p></li>
</ul>
</div>
<div id="transforming-single-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#transforming-single-interval-sets" class="anchor"></a>Transforming single interval sets</h3>
<ul>
<li><p>Adjust interval coordinates with <code><a href="../reference/bed_slop.html">bed_slop()</a></code> and <code><a href="../reference/bed_shift.html">bed_shift()</a></code>, and create new flanking intervals with <code><a href="../reference/bed_flank.html">bed_flank()</a></code>.</p></li>
<li><p>Combine nearby intervals with <code><a href="../reference/bed_merge.html">bed_merge()</a></code> and identify nearby intervals with <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>.</p></li>
<li><p>Generate intervals not covered by a query with <code><a href="../reference/bed_complement.html">bed_complement()</a></code>.</p></li>
<li><p>Order intervals with <code><a href="https://dplyr.tidyverse.org/reference/arrange.html">dplyr::arrange()</a></code>.</p></li>
</ul>
</div>
<div id="comparing-multiple-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-multiple-interval-sets" class="anchor"></a>Comparing multiple interval sets</h3>
<ul>
<li><p>Find overlaps between sets of intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>.</p></li>
<li><p>Apply functions to overlapping sets of intervals with <code><a href="../reference/bed_map.html">bed_map()</a></code>.</p></li>
<li><p>Remove intervals based on overlaps with <code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>.</p></li>
<li><p>Find overlapping intervals within a window with <code><a href="../reference/bed_window.html">bed_window()</a></code>.</p></li>
<li><p>Find closest intervals independent of overlaps with <code><a href="../reference/bed_closest.html">bed_closest()</a></code>.</p></li>
</ul>
</div>
<div id="randomizing-intervals" class="section level3">
<h3 class="hasAnchor">
<a href="#randomizing-intervals" class="anchor"></a>Randomizing intervals</h3>
<ul>
<li><p>Generate random intervals with <code><a href="../reference/bed_random.html">bed_random()</a></code>.</p></li>
<li><p>Shuffle the coordinates of intervals with <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code>.</p></li>
<li><p>Sample input intervals with <code><a href="https://dplyr.tidyverse.org/reference/sample.html">dplyr::sample_n()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/sample.html">dplyr::sample_frac()</a></code>.</p></li>
</ul>
</div>
<div id="interval-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-statistics" class="anchor"></a>Interval statistics</h3>
<ul>
<li><p>Calculate significance of overlaps between sets of intervals with <code><a href="../reference/bed_fisher.html">bed_fisher()</a></code> and <code><a href="../reference/bed_projection.html">bed_projection()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between sets of intervals with <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>.</p></li>
</ul>
</div>
<div id="utilities" class="section level3">
<h3 class="hasAnchor">
<a href="#utilities" class="anchor"></a>Utilities</h3>
<ul>
<li><p>Create features from BED12 files with <code><a href="../reference/create_introns.html">create_introns()</a></code>, <code><a href="../reference/create_tss.html">create_tss()</a></code>, <code><a href="../reference/create_utrs5.html">create_utrs5()</a></code>, and <code><a href="../reference/create_utrs3.html">create_utrs3()</a></code>.</p></li>
<li><p>Visualize the actions of valr functions with <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code>.</p></li>
<li><p>Constrain intervals to a genome reference with <code><a href="../reference/bound_intervals.html">bound_intervals()</a></code>.</p></li>
<li><p>Subdivide intervals with <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code>.</p></li>
<li><p>Convert BED12 to BED6 format with <code><a href="../reference/bed12_to_exons.html">bed12_to_exons()</a></code>.</p></li>
<li><p>Calculate spacing between intervals with <code><a href="../reference/interval_spacing.html">interval_spacing()</a></code>.</p></li>
</ul>
</div>
</div>
<div id="related-work" class="section level2">
<h2 class="hasAnchor">
<a href="#related-work" class="anchor"></a>Related work</h2>
<ul>
<li><p>Command-line tools <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> and <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://CRAN.R-project.org/package=bedr">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a> and <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
